pkg/px-resource-gateway/semaphore.go

```diff
@@ -0,0 +1,302 @@
+	s.heartBeatMutex.Lock()
+	defer s.heartBeatMutex.Unlock()
+
+	logrus.Debugf("Received KeepAlive request for node %v", clientId)
+	s.heartbeats[clientId] = time.Now()
+}
+
+func (s *semaphorePriorityQueue) fetchAvailableLock() uint32 {
+	if len(s.availableLocks) == 0 {
+		return NULL_LOCK
+	}
+	lockId := s.availableLocks[0]
+	s.availableLocks = s.availableLocks[1:]
+	return lockId
+}
+
+func (s *semaphorePriorityQueue) tryLock(clientId string) bool {
+	// check if the node is at the front of the queue
+	nextResouceId, priority := s.priorityQ.Front()
+	if nextResouceId == "" {
+		panic("Queue is empty")
+	}
+	logrus.Debugf("Next resource in queue: %v", nextResouceId)
+	if nextResouceId != clientId {
+		return false
+	}
+	// check if any lock is available
+	lockId := s.fetchAvailableLock()
+	if lockId == NULL_LOCK {
+		return false
+	}
+
+	s.locks[clientId] = lockId
+	s.priorityQ.Dequeue(priority) // cleanup
+
+	return true
+}
+
+func (s *semaphorePriorityQueue) updateConfigMap() error {
+
+	isUpdateRequired := func(map1, map2 map[string]string) bool {
+		if len(map1) != len(map2) {
+			return true
+		}
+		for key, value := range map1 {
+			if val, ok := map2[key]; !ok || val != value {
+				return true
+			}
+		}
+		return false
+	}
+
+	data := map[string]string{} // reset the data
+	for clientId, lockId := range s.locks {
+		data[clientId] = fmt.Sprintf("%d", lockId)
+	}
+
+	if !isUpdateRequired(data, s.configMap.Data) {
+		return nil
+	}
+
+	logrus.Infof("Updating configmap: %v", s.locks)
+	configMap, err := core.Instance().UpdateConfigMap(s.configMap)
+	if err != nil {
+		return err
+	}
+	s.configMap = configMap
+	return nil
+}
+
+func (s *semaphorePriorityQueue) configMapUpdateRoutine() {
+	ticker := time.NewTicker(ConfigMapUpdateInterval)
+	defer ticker.Stop()
+	for range ticker.C {
+		err := s.updateConfigMap()
+		if err != nil {
+			panic(err)
+		}
+		// Signal other goroutines that the work is done
+		close(s.configMapUpdateDone)
+		s.configMapUpdateDone = make(chan struct{})
+	}
+}
+
+func (s *semaphorePriorityQueue) cleanupDeadNodes() {
+	deadNodes := []string{}
+	removedNode := map[string]string{}
+
+	s.mutex.Lock()
+	s.heartBeatMutex.Lock()
+	defer func() {
+		s.heartBeatMutex.Unlock()
+		s.mutex.Unlock()
+		if len(deadNodes) != 0 {
+			<-s.configMapUpdateDone
+		}
+	}()
+
+	for clientId, lastHeartbeat := range s.heartbeats {
+		if time.Since(lastHeartbeat) > DeadNodeTimeout {
+			delete(s.heartbeats, clientId) // remove the resource footprint
+			deadNodes = append(deadNodes, clientId)
+		}
+	}
+	if len(deadNodes) == 0 {
+		return
+	}
+	logrus.Warnf("Cleaning up dead nodes: %v", deadNodes)
+
+	for _, clientId := range deadNodes {
+		if lockId, ok := s.locks[clientId]; ok {
+			s.release(clientId, lockId)
+			removedNode[clientId] = ""
+		}
+	}
+
+	for _, clientId := range deadNodes {
+		if _, ok := removedNode[clientId]; ok { // already removed
+			continue
+		}
+		s.priorityQ.Remove(clientId)
```

Comment:
I am missing something here, why not remove a dead node unconditionally from the priorityQueue if it was part of s.locks ?